src/main/java/com/mongodb/kafka/connect/source/schema/AvroSchema.java

```diff
@@ -0,0 +1,275 @@
+  }
+
+  static Schema createSchema(
+      final org.apache.avro.Schema avroSchema,
+      final boolean isOptional,
+      final Object defaultValue,
+      final Context context) {
+    SchemaBuilder builder;
+    switch (avroSchema.getType()) {
+      case RECORD:
+        SchemaBuilder structBuilder = SchemaBuilder.struct();
+        context.schemaCache.put(avroSchema, structBuilder);
+        structBuilder.name(avroSchema.getName());
+        avroSchema
+            .getFields()
+            .forEach(
+                f -> {
+                  if (context.schemaCache.containsKey(f.schema())) {
+                    context.detectedCycles.add(f.schema());
+                    structBuilder.field(f.name(), context.schemaCache.get(f.schema()));
+                  } else {
+                    Schema fieldSchema = createSchema(f.schema(), false, f.defaultVal(), context);
+                    structBuilder.field(f.name(), fieldSchema);
+                  }
+                });
+        builder = structBuilder;
+        break;
+      case MAP:
+        builder =
+            SchemaBuilder.map(
+                Schema.STRING_SCHEMA,
+                createSchemaCheckCycles(avroSchema.getValueType(), defaultValue, context));
+        break;
+      case ARRAY:
+        builder =
+            SchemaBuilder.array(
+                createSchemaCheckCycles(avroSchema.getElementType(), defaultValue, context));
+        break;
+      case STRING:
+        builder = SchemaBuilder.string();
+        break;
+      case BYTES:
+      case FIXED:
+        builder = SchemaBuilder.bytes();
+        break;
+      case INT:
+        builder = SchemaBuilder.int32();
+        break;
+      case LONG:
+        builder = SchemaBuilder.int64();
+        break;
+      case FLOAT:
+        builder = SchemaBuilder.float32();
+        break;
+      case DOUBLE:
+        builder = SchemaBuilder.float64();
+        break;
+      case BOOLEAN:
+        builder = SchemaBuilder.bool();
+        break;
+      case UNION:
+        Optional<org.apache.avro.Schema> optionalSchema =
+            avroSchema.getTypes().stream()
+                .filter(s -> s.getType() != org.apache.avro.Schema.Type.NULL)
+                .findFirst();
+        if (optionalSchema.isPresent()) {
+          return createSchema(optionalSchema.get(), true, context);
+        }
+        throw new IllegalStateException();
+      case NULL:
+      case ENUM:
+      default:
+        throw new IllegalStateException();
+    }
+
+    if (isOptional) {
+      builder.optional();
+    }
+
+    if (defaultValue != null) {
+      builder.defaultValue(processDefaultValue(builder, defaultValue));
+    }
+
+    if (!context.detectedCycles.contains(avroSchema)) {
+      context.schemaCache.remove(avroSchema);
+    }
+
+    return builder.build();
+  }
+
+  static Object processDefaultValue(final SchemaBuilder schemaBuilder, final Object value) {
+    if (schemaBuilder.type() == Type.STRUCT) {
```

Comment:
There don't appear to be any unit tests where we enter this conditional.  Can we add one?

Reply:
Added